### DISCUSSION TITLE
My function require so much gas

### DISCUSSION THREAD

**Discussion by FurkanSezal** [zkSync-Community-Hub/zksync-developers#111] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I have a proxy nft marketplace contract. Its works perfectly on evm chains. I successfully deployed to zk sync testnet but my contract function requires so much gas. It requires so little on evm chain. What can cause this? 

### What is the specific issue or error you're encountering?

![Ekran görüntüsü 2023-08-21 132416](https://github.com/zkSync-Community-Hub/zkync-developers/assets/102124127/c4b588d7-a224-4d14-b523-254f2b479b72)

### Can you share the error messages or logs you're receiving, if any?

![Ekran görüntüsü 2023-08-21 132416](https://github.com/zkSync-Community-Hub/zkync-developers/assets/102124127/c36154a5-fae3-478a-8617-7dc96bf44262)

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

no 

### Have you tried to isolate the problem, and if so, what were the results?

no

### What steps have you already taken to try to resolve the issue?

I though try to give manual gasLimit but its not recommended

**Comment by githubdoramon** [comment]
Can you please share the contract address and code?

**Reply by FurkanSezal** [reply]
Unfortunately i cant share the code its a private project. But its working on evm and its also cheap on evm. 
I tried this zksolc optimizer wasnt enabled. But it didnt help.  I know its hard say something without contract code. Maybe you can make suggestions. Because its working on evm ( also on ethereum mainnet ) i dont think its causing from contract.   

solidity: {
    version: "0.8.18",
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },
  zksolc: {
    version: "latest",
    compilerSource: "binary",
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },

**Reply by FurkanSezal** [reply]
I will be comfortable share the contract in private is it possible?

**Reply by githubdoramon** [reply]
@FurkanSezal send your contact to support@zksync.io so I can reach out to you privately please.

**Reply by githubdoramon** [reply]
@FurkanSezal got your code. Can you let me know which function are you calling in this example?

**Reply by FurkanSezal** [reply]
buy, cancelOrder ,acceptGlobalBid , acceptBid all give the same error. 
![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/102124127/3e3bdbb6-f230-4af5-95c4-d16f1bf72817)

**Reply by FurkanSezal** [reply]
On bit5.sol contract. That is the main marketplace contract

**Reply by FurkanSezal** [reply]
@githubdoramon is there any news for me?

**Reply by githubdoramon** [reply]
Still looking into it @FurkanSezal. Sorry for the delay, but team is quite packed with a lot of things. We will get back to you for sure.

**Reply by FurkanSezal** [reply]
Thank you for the reply. Looking forward to solution

**Reply by mm-zk** [reply]
@FurkanSezal - do you have it deployed somewhere on testnet? (if yes - can you share the transaction data that is causing such a huge fee?)

**Reply by FurkanSezal** [reply]
Yes its on zksync testnet. https://test.bit5.com/collection/0x118/akitavaxtest go head test it out. Try to buy some of the nfts

**Reply by FurkanSezal** [reply]
@githubdoramon good day any news?

**Reply by mm-zk** [reply]
Hey - I tried to use the page that you sent - but I got a bunch of errors (including '
err:TypeError: Cannot mix BigInt and other types, use explicit conversions') and 'fractional component exceeds decimals
' and then finally I got 'missing gas limit' in my metamask.

Can you provide the example transaction payload ?

**Reply by mm-zk** [reply, marked as answer]
@FurkanSezal - we discovered an issue with one of our APIs (eth_feeHistory) - that might have caused some wallets to set too high gas PRICE when estimating the cost of the transaction.

We've just rolled out the fix to both testnet and mainnet. Can you see if your issue still persists?
(and can you check what was the gas limit and gas price for that 1ETH transaction?)

**Comment by caydiverlarachel** [comment]
The increased gas consumption on zkSync compared to EVM chains could be due to differences in the underlying architecture and transaction processing mechanisms. zkSync relies on zero-knowledge proofs and off-chain computations to achieve scalability, which may result in higher gas costs for certain types of transactions. Additionally, the specific implementation of your contract and its interactions with zkSync may contribute to the gas usage discrepancy. Consider optimizing your contract code and transaction logic to minimize gas consumption on zkSync while maintaining functionality. Additionally, reaching out to the zkSync community or developers for insights and assistance could be beneficial in resolving this issue.

**Comment by pobertester** [comment]
Environment: Testnet
Issue: High gas consumption for contract functions on zkSync compared to EVM chains
Potential Cause: Differences in gas calculation and execution between zkSync and EVM chains
Recommendation: Review contract logic and interactions with zkSync to identify potential optimizations for gas usage
Differences in gas calculation and execution between zkSync and EVM chains may lead to higher gas consumption for contract functions on zkSync. Reviewing contract logic and interactions with zkSync can help identify potential optimizations for gas usage.

**Comment by dariuscandfield** [comment]
It's possible that the gas cost difference is due to differences in the underlying execution environment between EVM chains and zkSync. zkSync operates on a layer 2 scaling solution, which may have different gas costs for certain operations compared to layer 1 EVM chains. Reviewing your contract's operations and optimizing gas usage where possible may help mitigate this issue. Additionally, considering zkSync-specific optimizations and gas-efficient patterns could further improve performance on the zkSync network.

**Comment by marcellusjustin** [comment]
It's possible that the gas requirements for your contract function on the zk sync testnet are higher due to the different architecture of the network. You may need to optimize your contract code or adjust the gas limit to ensure successful execution on the testnet.

**Comment by cioprogephen** [comment]
The user has a proxy NFT marketplace contract that works perfectly on EVM chains but requires significantly more gas on the zkSync Testnet. The user has provided a screenshot of an error message but no specific details or relevant portions of their contract code. The user has not made any recent changes to the contract before encountering this issue and is not interacting with any external libraries or contracts. The user has not tried to isolate the problem or taken any steps to try to resolve the issue, except for manually setting the gasLimit, which is not recommended. No repo link is provided.
